mm/vm_map/vm_map.c

```diff
@@ -0,0 +1,296 @@
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <nuttx/queue.h>
+#include <nuttx/sched.h>
+#include <nuttx/semaphore.h>
+#include <nuttx/mm/vm_map.h>
+#include <nuttx/kmalloc.h>
+#include <sys/mman.h>
+#include <assert.h>
+
+#if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static bool vaddr_in_area(FAR const void *addr, FAR const void *start,
+                          size_t length)
+{
+  uintptr_t u_addr = (uintptr_t)addr;
+  uintptr_t u_start = (uintptr_t)start;
+  return (u_addr >= u_start) && (u_addr < u_start + length);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: vm_map_initialize
+ *
+ * Description:
+ *   Allocates a task group specific vm_map stucture. Called when the group
+ *   is initialized
+ *
+ ****************************************************************************/
+
+void vm_map_initialize(FAR struct vm_map_s *mm)
+{
+  memset(mm, 0, sizeof(struct vm_map_s));
+  sq_init(&mm->vm_map_sq);
+  nxsem_init(&mm->vm_map_sem, 0, 1);
+}
+
+/****************************************************************************
+ * Name: vm_map_destroy
+ *
+ * Description:
+ *   De-allocates a task group specific vm_map stucture and the vm_map_sem
+ *
+ ****************************************************************************/
+
+void vm_map_destroy(FAR struct vm_map_s *mm)
+{
+  FAR struct vm_map_entry_s *map =
+    (struct vm_map_entry_s *)sq_peek(&mm->vm_map_sq);
+  while (map)
+    {
+      munmap((void *)map->vaddr, map->length);
```

Comment:
This does not work if the TCB is removed from the task lisk (RTR head != currently running task). This happens when a task exits, or another process terminates a task (in flat mode via task_delete)

Reply:
Maybe it is worth fixing this issue still before merging even to our main. We are not that much in a hurry with this.  I will work on a solution.